### PR TITLE
[DataGrid] Fix `sortModel` and `filterModel` resetting when columns change

### PR DESCRIPTION
**Pull request by alexgonch**
This PR fixes the problem raised in the #9204 issue by implementing the suggestion of @cherniavskii 

By replacing useEffect with useLayoutEffect, the `setSortModel` and `setFilterModel` methods are now called before the `handleColumnsChange` runs. This ensures that when both the columns and the respective model(s) change simultaneously, the DataGrid won't fire the onSortModelChange and onFilterModelChange events with erroneously empty models as their parameters.

This change fixes the problems I've had in my project, but I haven't tested it beyond that.

### CODE REVIEW

File: packages/grid/x-data-grid/src/hooks/features/filter/useGridFilter.tsx [hunk 1]
```diff
@@ -485,7 +485,7 @@ export const useGridFilter = (
   /**
    * EFFECTS
    */
-  React.useEffect(() => {
+  React.useLayoutEffect(() => {
```
Comment (flaviendelangle) [inline packages/grid/x-data-grid/src/hooks/features/filter/useGridFilter.tsx:488]: You would need to use `import { unstable_useEnhancedEffect as useEnhancedEffect } from '@mui/utils';` in order for SSR to work

**Review by cherniavskii — APPROVED** [review]
Thanks!
I also added unit tests for both filtering and sorting 👍

File: packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx [hunk 2]
```diff
@@ -1248,4 +1249,76 @@ describe('<DataGrid /> - Filter', () => {
...
+        </div>
+      );
+    }
+    const { setProps } = render(<Demo rows={rows} />);
+    expect(getColumnValues(1)).to.deep.equal(['Puma']);
+
+    setProps({
+      columns: [{ field: 'id' }],
+      filterModel: {
+        items: [{ field: 'id', operator: 'equals', value: '1' }],
```
Comment (MBilalShafi) [inline packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx:1318]: As the `filterModel` prop is updated, technically it's changed, shouldn't this emit a `filterModelChange` here? Or we only emit it once it's internally changed. (Same with `sortModel`)

Comment (cherniavskii) [reply packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx:1318]: Good question!
In controlled mode, we don't call onChange handler if the new value is the same as the prop:
https://github.com/mui/mui-x/blob/a7b87a4be7c98dd5f28a0a51fce49e692af61c88/packages/grid/x-data-grid/src/hooks/core/useGridStateInitialization.ts#L61

https://github.com/mui/mui-x/blob/a7b87a4be7c98dd5f28a0a51fce49e692af61c88/packages/grid/x-data-grid/src/hooks/core/useGridStateInitialization.ts#L97-L103

Comment (jonnylink) [reply packages/grid/x-data-grid/src/tests/filtering.DataGrid.test.tsx:1318]: Just a guess, but could this by we are still experiencing an issue? Refresh once and the filters/sorts are missing. Refresh again and they are back.

**Comment by cherniavskii** [comment]
Hey @jonnylink 
Could you please open a new issue with a minimal reproduction example?
Thank you!

**Comment by jonnylink** [comment]
@cherniavskii 
Sure thing. I'll do that as soon as I have some time. Under a few tight deadlines. Our setup might not be super easy to reproduce (we use an IndexedDB persistent TanStack cache). 

Seeing as this bug fix got us halfway there, I'm fairly confident it's related.

**Comment by fabianboerner** [comment]
this is not fixed in version 5?

**Comment by cherniavskii** [comment]
@fabianboerner No, the latest stable version is v6 and we're going to release the v7 very soon.
We encourage you to migrate to v6.

**Pull request merged**